server/src/sql-tools/schema-diff.ts

```diff
@@ -0,0 +1,449 @@
+  if (source.default === target.default) {
+    return true;
+  }
+
+  if (source.default === undefined || target.default === undefined) {
+    return false;
+  }
+
+  if (
+    withTypeCast(source.default, getColumnType(source)) === target.default ||
+    source.default === withTypeCast(target.default, getColumnType(target))
+  ) {
+    return true;
+  }
+
+  return false;
+};
+
+/**
+ * Compute the difference between two database schemas
+ */
+export const schemaDiff = (
+  source: DatabaseSchema,
+  target: DatabaseSchema,
+  options: { ignoreExtraTables?: boolean } = {},
+) => {
+  const items = diffTables(source.tables, target.tables, {
+    ignoreExtraTables: options.ignoreExtraTables ?? true,
+  });
+
+  return {
+    items,
+    asSql: (options?: SchemaDiffToSqlOptions) => schemaDiffToSql(items, options),
+  };
+};
+
+export const diffTables = (
+  sources: DatabaseTable[],
+  targets: DatabaseTable[],
+  options: { ignoreExtraTables: boolean },
+) => {
+  const items: SchemaDiff[] = [];
+  const sourceMap = Object.fromEntries(sources.map((table) => [table.name, table]));
+  const targetMap = Object.fromEntries(targets.map((table) => [table.name, table]));
+  const keys = new Set([...Object.keys(sourceMap), ...Object.keys(targetMap)]);
+
+  for (const key of keys) {
+    if (options.ignoreExtraTables && !sourceMap[key]) {
+      continue;
+    }
+    items.push(...diffTable(sourceMap[key], targetMap[key]));
+  }
+
+  return items;
+};
+
+const diffTable = (source?: DatabaseTable, target?: DatabaseTable): SchemaDiff[] => {
+  if (isSynchronizeDisabled(source, target)) {
+    return [];
+  }
+
+  if (source && !target) {
+    return [
+      {
+        type: 'table.create',
+        tableName: source.name,
+        columns: Object.values(source.columns),
+        reason: Reason.MissingInTarget,
+      },
+      ...diffIndexes(source.indexes, []),
+      // TODO merge constraints into table create record when possible
+      ...diffConstraints(source.constraints, []),
+    ];
+  }
+
+  if (!source && target) {
+    return [
+      {
+        type: 'table.drop',
+        tableName: target.name,
+        reason: Reason.MissingInSource,
+      },
+    ];
+  }
+
+  if (!source || !target) {
+    return [];
+  }
+
+  return [
+    ...diffColumns(source.columns, target.columns),
+    ...diffConstraints(source.constraints, target.constraints),
+    ...diffIndexes(source.indexes, target.indexes),
+  ];
+};
+
+const diffColumns = (sources: DatabaseColumn[], targets: DatabaseColumn[]): SchemaDiff[] => {
+  const items: SchemaDiff[] = [];
+  const sourceMap = Object.fromEntries(sources.map((column) => [column.name, column]));
+  const targetMap = Object.fromEntries(targets.map((column) => [column.name, column]));
+  const keys = new Set([...Object.keys(sourceMap), ...Object.keys(targetMap)]);
+
+  for (const key of keys) {
+    items.push(...diffColumn(sourceMap[key], targetMap[key]));
+  }
+
```

Comment:
This happens in a couple of places here. Thoughts on extracting it into a helper function?